zds/notification/managers.py

```diff
@@ -0,0 +1,191 @@
+                    content_type__pk=content_type.pk,
+                    user=user, is_active=is_active,
+                    by_email=by_email)
+        except ObjectDoesNotExist:
+            existing = None
+        return existing
+
+    def get_or_create_active(self, user, content_object):
+        """
+        Gets (or create if it doesn't exist) the subscription for the content object given.
+        :param user: concerned user.
+        :param content_object: Generic content concerned.
+        :return:
+        """
+        content_type = ContentType.objects.get_for_model(content_object)
+        try:
+            subscription = self.get(
+                object_id=content_object.pk,
+                content_type__pk=content_type.pk,
+                user=user)
+            if not subscription.is_active:
+                subscription.activate()
+        except ObjectDoesNotExist:
+            subscription = self.model(user=user, content_object=content_object)
+            subscription.save()
+
+        return subscription
+
+    def get_subscriptions(self, content_object, is_active=True):
+        """
+        Gets subscriptions of the content object.
+        :param content_object: Generic content concerned.
+        :param is_active: Boolean to know if we want a subscription active or not.
+        :return:
+        """
+        content_type = ContentType.objects.get_for_model(content_object)
+        return self.filter(object_id=content_object.pk,
+                           content_type__pk=content_type.pk,
+                           is_active=is_active)
+
+    def get_subscribers(self, content_object, only_by_email=False):
+        """
+        Gets all subscribers of a content object.
+        :param content_object: Generic content concerned.
+        :param only_by_email: Boolean to know if we want a subscription for email or not.
+        :return:
+        """
+        content_type = ContentType.objects.get_for_model(content_object)
+        if only_by_email:
+            # if I'm only interested by the email subscription
+            subscription_list = self.filter(
+                object_id=content_object.pk,
+                content_type__pk=content_type.pk,
+                by_email=True)
+        else:
+            subscription_list = self.filter(
+                object_id=content_object.pk,
+                content_type__pk=content_type.pk)
+
+        return [subscription.user for subscription in subscription_list]
+
+    def get_objects_followed_by(self, user):
+        """
+        Gets objects followed by the given user.
+        :param user: concerned user.
+        :return: All objects followed by given user.
+        """
+        subscription_list = self.filter(user=user, is_active=True) \
+            .order_by('last_notification__pubdate')
+
+        return [subscription.content_object for subscription in subscription_list]
+
+    def toggle_follow(self, content_object, user=None, by_email=None):
```

Comment:
Je verrais aussi plutôt =False que =None, avec `if not user:` à la place de `if user is None`.